Documentation/proxy.md

```diff
@@ -6,9 +6,11 @@ etcd currently supports two proxy modes: `readwrite` and `readonly`. The default
 
 The proxy will shuffle the list of cluster members periodically to avoid sending all connections to a single member.
 
-The member list used by proxy consists of all client URLs advertised within the cluster, as specified in each members' `-advertise-client-urls` flag. If this flag is set incorrectly, requests sent to the proxy are forwarded to wrong addresses and then fail. Including URLs in the `-advertise-client-urls` flag that point to the proxy itself, e.g. http://localhost:2379, is even more problematic as it will cause loops, because the proxy keeps trying to forward requests to itself until its resources (memory, file descriptors) are eventually depleted. The fix for this problem is to restart etcd member with correct `-advertise-client-urls` flag. After client URLs list in proxy is recalculated, which happens every 30 seconds, requests will be forwarded correctly.
+The member list used by proxy consists of all client URLs advertised within the cluster, as specified in each member's `advertise-client-urls` flag that represents client URLs to advertise to the public.
 
-You should start the cluster machines first and then start the proxy as the proxy gets its list from reachable machines. If the proxy starts and is unable to reach a machine it will refresh aggressively. Once the machine list is non-empty the proxy will refresh at the original duration.
+etcd proxy first tries to get its endpoints(peerURLs) from `discovery` or `discovery-srv` flag, otherwise from `initial-cluster` flag. And then gets clientURLs from the first reachable member from the peerURLs. And this clientURLs is set by the flag `advertise-client-urls`. And proxy connects to this clientURLs every 30-second in order to refresh its endpoints. This is why we need both `initial-cluster` and `advertise-client-urls` flags for every member in cluster, and proxy does not need `advertise-client-urls` flag. If these flags aren't set correctly, requests to proxy would be forwarded to wrong addresses. Having URLs in `advertise-client-urls` that points to the proxy itself can cause proxy loops forwarding requests to itself until resource depletion. In this case, you need to restart etcd member with correct `advertise-client-urls` flag.
+
+You shoul  start the cluster machines first and then start the proxy as the proxy gets its list from reachable machines. If the proxy starts and is unable to reach a machine it will refresh aggressively. Once the machine list is non-empty the proxy will refresh at the original duration.
```

Comment:
Always start the first-order etcd cluster members first, then any proxies. A proxy must be able to reach the cluster members to retrieve its configuration, and will attempt connections somewhat aggressively in the absence of such a channel. Starting the members before any proxy ensures the proxy can discover the client URLs when it later starts.